### DISCUSSION TITLE
Production costs not different between levels

### DISCUSSION THREAD

**Discussion by dipplestix** [yasserfarouk/scml#36] [one-shot-qa] [answered]
Running the following code
```
agent_types = [RandomOneShotAgent, SyncRandomOneShotAgent, GreedyOneShotAgent, GreedySingleAgreementAgent]

val_0 = 0
val_1 = 0
val_2 = 0
val_3 = 3
val_4 = 4
val_5 = 5
n = 1000
for i in range(n):
    world = SCML2020OneShotWorld(
        **SCML2020OneShotWorld.generate(
            agent_types=agent_types,
            n_steps=50
        ),
        construct_graphs=True,
    )
    if i == 0:
        print(world.agent_profiles)

    val_0 += world.agent_profiles[list(world.agent_profiles.keys())[0]].cost
    val_1 += world.agent_profiles[list(world.agent_profiles.keys())[1]].cost
    val_2 += world.agent_profiles[list(world.agent_profiles.keys())[2]].cost
    val_3 += world.agent_profiles[list(world.agent_profiles.keys())[3]].cost
    val_4 += world.agent_profiles[list(world.agent_profiles.keys())[4]].cost
    val_5 += world.agent_profiles[list(world.agent_profiles.keys())[5]].cost
print(np.array([val_0, val_1, val_2, val_3, val_4, val_5])/n)
```

yields
```
defaultdict(<class 'list'>, {'00SyR@0': OneShotProfile(cost=6, input_product=0, n_lines=10, shortfall_penalty_mean=0.7077609276422603, disposal_cost_mean=0.8855099889110551, shortfall_penalty_dev=0.05173409710349367, disposal_cost_dev=0.04921604085230723), '01GSA@0': OneShotProfile(cost=4, input_product=0, n_lines=10, shortfall_penalty_mean=0.5263103223795285, disposal_cost_mean=0.5201459953792971, shortfall_penalty_dev=0.044325413557489104, disposal_cost_dev=0.008711963878956994), '02Ran@0': OneShotProfile(cost=7, input_product=0, n_lines=10, shortfall_penalty_mean=0.7358165454416139, disposal_cost_mean=0.8990585669339071, shortfall_penalty_dev=0.06862436053457982, disposal_cost_dev=0.05497332260213158), '03GSA@1': OneShotProfile(cost=8, input_product=1, n_lines=10, shortfall_penalty_mean=0.6088231183711137, disposal_cost_mean=0.6677210453076889, shortfall_penalty_dev=0.05990315501649243, disposal_cost_dev=0.053014124340221), '04GSA@1': OneShotProfile(cost=9, input_product=1, n_lines=10, shortfall_penalty_mean=0.8639006169834454, disposal_cost_mean=0.677402915478307, shortfall_penalty_dev=0.04869479899135038, disposal_cost_dev=0.043469001032185775), '05SyR@1': OneShotProfile(cost=9, input_product=1, n_lines=10, shortfall_penalty_mean=0.9355877619447468, disposal_cost_mean=0.9469732340406642, shortfall_penalty_dev=0.05625999398377608, disposal_cost_dev=0.028563629972440628)})

[5.404 5.578 5.632 5.645 5.405 5.653]
```

According to the document, the factories with input_product=1 should have double the average cost as the ones at input_product=0. Is the document wrong or is this unintended behavior?

**Comment by yasserfarouk** [comment, marked as answer]
Thank you for noticing this issue.

I think what you are referring to in the description document is this line:

> In all SCML tracks, production becomes more and more costly, on average,
the closer a product comes to being finished.

In Tables 2 and 3 we specify how exactly is this supposed to be implemented. Firstly, the *mean* production cost of each **product** should depend (in an unspecified way) on the level and should be between 1 and 10 (that is $m_l$ in Table 2). After that a production cost **per agent** is sampled from a uniform distribution that ranges from this number to one order of magnitude higher (Table 3). 

Firstly, in one-shot we have a single intermediate product. Secondly, this large range of production costs will mostly mask the difference in the *mean* used for sampling. This is by design and we want the factories to have a wide variety of ufuns which is helped by having a large range for production costs.

**update after checking issue #38 **

I think the description document should be modified to make it clearer that what increases with the level is not the **empirical average** but the **location** of the distribution used for sampling. This means that with the large *scale* shown in Table 3, you should not be able to see a difference in the mean with a small sample of say 6 agents. To see a difference, may be we need to have hundreds of agents in both levels.

**Reply by dipplestix** [reply]
Why is "l x U[1, 10]" unspecified? Isn't this just the level (1 or 2) times the uniform rv?

Also, I never see an `agent.awi.profile.cost` or `agent.ufun.production_cost` of value greater than 10, so when does the sampling mentioned in table 3 occur? (these values are always equal to each other)

**Reply by dipplestix** [reply]
```
print(agent.ufun.production_cost)
print(agent.awi.level)
print(agent.awi.profile.cost)

10
1
10

world.step()

True

print(agent.ufun.production_cost)
print(agent.awi.level)
print(agent.awi.profile.cost)

10
1
10
```

**Reply by yasserfarouk** [reply]
> Why is "l x U[1, 10]" unspecified? Isn't this just the level (1 or 2) times the uniform rv?

It just says that we sample two numbers uniformly then multiply them by *l*. That does not guarantee that the resulting number will be higher for the higher level. For example, if the two numbers were [5, 1] then the resulting m_l values will be 5, 2. If we average over a large number of simulations, we will see that the production cost is higher in the second layer though and almost double but in your example, we are considering a single sample (i.e. a single world).


> Also, I never see an agent.awi.profile.cost or agent.ufun.production_cost of value greater than 10, so when does the sampling mentioned in table 3 occur?

The sampling in Table 3 happens here:
```
        production_costs = make_array(production_costs, n_agents, dtype=int)
```

which will always sample a number between 1 and 10. 

**After that**, we run the following to increase the cost of higher processes (achieving what is described in Table 2). This is the code that achieves the effect:

```
        for i, (f, l) in enumerate(zip(first_agent, last_agent)):
            process_of_agent[f:l] = i
            if cost_increases_with_level:
                production_costs[f:l] = np.round(
                    production_costs[f:l] * (i + 1)  # math.sqrt(i + 1)
                ).astype(int)
```

Also, the production cost does not change over time. It is fixed for the whole simulation time. Everything in the profile is.

In summary, to see a difference between the layers in oneshot, you need to sample several worlds and average over them. Moreover, you need to have say 100 agents per layer in each world because of the large variation in the sampling procedure in Table 3.

I will check it again and confirm that in the limit the document is accurate. Nevertheless, I now think that we should not have added this statement to the description.

Finally, why is this important to you? I mean the difference in production costs between different levels does not seem to be an important factor in devising a strategy for the agent. I am missing something? May be there is a deeper confusion here.

**Reply by dipplestix** [reply]
> > Why is "l x U[1, 10]" unspecified? Isn't this just the level (1 or 2) times the uniform rv?
> 
> It just says that we sample two numbers uniformly then multiply them by _l_. That does not guarantee that the resulting number will be higher for the higher level. For example, if the two numbers were [5, 1] then the resulting m_l values will be 5, 2. If we average over a large number of simulations, we will see that the production cost is higher in the second layer though and almost double but in your example, we are considering a single sample (i.e. a single world).

In my example at the start of the discussion, I average over 1000 worlds and get the 5.5 average for all agents
 
> **After that**, we run the following to increase the cost of higher processes (achieving what is described in Table 2). This is the code that achieves the effect:
> 
> ```
>         for i, (f, l) in enumerate(zip(first_agent, last_agent)):
>             process_of_agent[f:l] = i
>             if cost_increases_with_level:
>                 production_costs[f:l] = np.round(
>                     production_costs[f:l] * (i + 1)  # math.sqrt(i + 1)
>                 ).astype(int)
> ```
> 
> Also, the production cost does not change over time. It is fixed for the whole simulation time. Everything in the profile is.

Same here, the sample over the different worlds never had a value over 10 in either of those two variables

> Finally, why is this important to you? I mean the difference in production costs between different levels does not seem to be an important factor in devising a strategy for the agent. I am missing something? May be there is a deeper confusion here.

The production cost isn't important to agent behavior, but the expected catalog price (and therefore value of the no agreement contract) is impacted significantly by it.

**Reply by yasserfarouk** [reply]
If you are averaging over 100 world runs, there is a problem. I guess it is a saturation effect somewhere. Will look into it and return to you.

Note that for oneshot, the increase in production cost per level has no effect on the cost of the intermediate product (as it is the output of the first process) but it will affect the catalog price of the final product as you indicated. Nevertheless, the margin of profit expected when the world is generated is taking that into account. 
Moreover catalog and trading prices are public knowledge so the agent should depend on them directly not on production costs of agents in the other layer.

I will comment back once I could confirm the issue and will resolve it to make the document match the implementation 

Thanks

**Reply by yasserfarouk** [reply]
I reproduced the issue. Will treat it as a bug. When resolved, the commit number will be posted here. thanks

**Reply by yasserfarouk** [reply]
The bug was fixed (see issue #38). Now the implementation matches the description. 

## Summary

Average production cost will increase with the product's number. Because of the large variation in production costs per level (an order of magnitude by design), this may not be apparent enough with small numbers of agents in a single world in some cases but averaging over say 100 worlds with say 10 agents per level, it will be apparent. 

To test this property run:

```
    from scml.oneshot.agents import GreedyOneShotAgent
    from scml.oneshot.world import SCML2020OneShotWorld

    NPROCESSES = 5
    costs = [[] for _ in range(NPROCESSES)]
    for _ in range(100):
        world = SCML2020OneShotWorld(
            **SCML2020OneShotWorld.generate(
                GreedyOneShotAgent,
                n_agents_per_process=10,
                n_processes=NPROCESSES,
            ),
            compact=True,
            no_logs=True,
        )
        for aid in world.agent_profiles.keys():
            if is_system_agent(aid):
                continue
            profile = world.agent_profiles[aid]
            costs[profile.input_product].append(profile.cost)
    mean_costs = [sum(_) / len(_) for _ in costs]
    assert all(
        [
            b > (0.5 * (i + 2) / (i + 1)) * a
            for i, (a, b) in enumerate(zip(mean_costs[:-1], mean_costs[1:]))
        ]
    ), f"non-ascending costs {mean_costs}"
```
